annotations/relations.py

```diff
@@ -556,3 +739,57 @@ def update_template(template, template_data, part_data_list):
             part.save()
 
     return template
+
+
+def clean_terminal_nodes(self):
+    value = self.cleaned_data.get('terminal_nodes')
+    if not value:
+        return value
+    
+    try:
+        # Parse the terminal nodes, validating format
+        parsed_nodes = [node.strip() for node in value.split(',')]
+        
+        # Get relation node values for validation
+        first_node_type = self.cleaned_data.get('first_node_type')
+        first_node_value = self.cleaned_data.get('first_node_value')
+        second_node_type = self.cleaned_data.get('second_node_type')
+        second_node_value = self.cleaned_data.get('second_node_value')
+        third_node_type = self.cleaned_data.get('third_node_type')
+        third_node_value = self.cleaned_data.get('third_node_value')
+        
+        # Check if using relation nodes
+        using_relation_nodes = first_node_type and first_node_value and \
+                              second_node_type and second_node_value and \
+                              third_node_type and third_node_value
+        
+        # If using relation nodes, ensure terminal nodes correspond to the Node values
+        if using_relation_nodes:
+            expected_nodes = []
+            if str(first_node_type) == str(DefaultMapping.NODE):
+                expected_nodes.append(first_node_value)
+            if str(second_node_type) == str(DefaultMapping.NODE):
+                expected_nodes.append(second_node_value)
+            if str(third_node_type) == str(DefaultMapping.NODE):
+                expected_nodes.append(third_node_value)
+            
+            # Check that all expected nodes are in the terminal nodes list
+            for node in expected_nodes:
+                if node not in parsed_nodes:
+                    raise ValidationError(f"Terminal nodes must include all Node values. Missing: {node}")
+        
+        # Also validate against expression field if present
+        expression = self.cleaned_data.get('expression')
+        if expression:
+            # Extract node references from expression
+            formatter = Formatter()
+            refs = [key for _, key, _, _ in formatter.parse(expression) if key is not None]
+            
+            # Ensure refs from expression match terminal nodes
+            for ref in refs:
+                if '{' + ref + '}' in expression and ref not in parsed_nodes:
+                    raise ValidationError(f"Terminal nodes must include all nodes referenced in the expression. Missing: {ref}")
+        
+        return value
```

Comment:
same as before. terminal nodes and default mapping are very different. they are independent of each other.

Reply:
wait, isn't that pretty much the same code as validating the nodes?